### ISSUE TITLE
Name resolution not reconfigured for AAAA when IPv6 active only on VPN link

### ISSUE THREAD

**Issue by mj-harvey** [Tunnelblick/Tunnelblick#490]
Hi,

When the OS is configured only with an IPv4 address (and possibly only a link-local IPv6 address), macOS' name resolution is configured only to request A records.
If Tunnelblick is then connected to an OpenVPN service that assigns both IPv4 and IPv6 addresss to the client, the name resolution subsystem does not get reconfigured to also request AAAA records.
The result is that services that are only IPv6-reachable over the VPN link are not accessible by hostname, only by IP.

If the OS already has working IPv6 configured before the VPN connection is established, then name resolution is already configured to look up A and AAAA records, so things work as expected.

A similar problem is described at https://apple.stackexchange.com/questions/304215/how-to-add-aaaa-flag-ipv6-to-dns-resolver-configuration-on-sierra

**Comment by mj-harvey** [comment]
This is with 3.7.8beta01. Irrespective of the setting of "disable IPv6 (yun only)", the DNS config remains:

> $ scutil --dns
DNS configuration
resolver #1
  search domain[0] : lan
  nameserver[0] : 192.168.1.254
  if_index : 7 (en0)
  flags    : Request A records
  reach    : 0x00020002 (Reachable,Directly Reachable Address)
resolver #2
  domain   : local
  options  : mdns
  timeout  : 5
  flags    : Request A records
  reach    : 0x00000000 (Not Reachable)
  order    : 300000
resolver #3
  domain   : 254.169.in-addr.arpa
  options  : mdns
  timeout  : 5
  flags    : Request A records
  reach    : 0x00000000 (Not Reachable)
  order    : 300200
resolver #4
  domain   : 8.e.f.ip6.arpa
  options  : mdns
  timeout  : 5
  flags    : Request A records
  reach    : 0x00000000 (Not Reachable)
  order    : 300400
resolver #5
  domain   : 9.e.f.ip6.arpa
  options  : mdns
  timeout  : 5
  flags    : Request A records
  reach    : 0x00000000 (Not Reachable)
  order    : 300600
resolver #6
  domain   : a.e.f.ip6.arpa
  options  : mdns
  timeout  : 5
  flags    : Request A records
  reach    : 0x00000000 (Not Reachable)
  order    : 300800
resolver #7
  domain   : b.e.f.ip6.arpa
  options  : mdns
  timeout  : 5
  flags    : Request A records
  reach    : 0x00000000 (Not Reachable)
  order    : 301000
DNS configuration (for scoped queries)
resolver #1
  search domain[0] : lan
  nameserver[0] : 192.168.1.254
  if_index : 7 (en0)
  flags    : Scoped, Request A records
  reach    : 0x00020002 (Reachable,Directly Reachable Address)


IPv6 is working - a `host` or `ping6` of a AAAA-named host accessible only via the VPN tunnel is successful

**Comment by wido** [comment]
Running into the same problem with Tunnelblick

Our OpenVPN server config:

<pre>
tun-ipv6
push tun-ipv6
push "redirect-gateway ipv6";
push "route-ipv6 2000::/3";
push "dhcp-option DNS6 2a00:f10:ff04:153::53"
push "dhcp-option DNS6 2a00:f10:ff04:253::53"
</pre>

The problem however remains:

<pre>
resolver #1
  nameserver[0] : 192.168.4.1
  if_index : 6 (en0)
  flags    : Scoped, Request A records
  reach    : 0x00020002 (Reachable,Directly Reachable Address)
</pre>

Running with OpenVPN 2.4.6 and Tunnelblick 3.7.7 build 5150 (2018-09-29)

**Comment by wido** [comment]
I also tried manually by backporting the scripts from #452 but that didn't work either.

IPv6 routes are set, but DNS isn't set to query AAAA, thus IPv6 does not work properly.

**Comment by mj-harvey** [comment]
Same behaviour with the new macOS Mojave release, as well.

**Comment by wido** [comment]
It has something to do with *scutil* which needs to be informed about the system being IPv6 capable.

My MacOS experience is very limited (Linux guy), so I would not know where to look or what to do.

**Comment by mj-harvey** [comment]
Have tried the trial of the [Viscosity OpenVPN client](https://www.sparklabs.com/viscosity/), which works.  See below for the output of `scutil --dns` when using that. The delta is that  "Request AAAA records" has been added to "resolver #1".

```
$ scutil --dns
DNS configuration
resolver #1
  search domain[0] : lan
  nameserver[0] : 192.168.1.254
  if_index : 7 (en0)
  flags    : Request A records, Request AAAA records
  reach    : 0x00020002 (Reachable,Directly Reachable Address)
resolver #2
  domain   : local
  options  : mdns
  timeout  : 5
  flags    : Request A records, Request AAAA records
  reach    : 0x00000000 (Not Reachable)
  order    : 300000
resolver #3
  domain   : 254.169.in-addr.arpa
  options  : mdns
  timeout  : 5
  flags    : Request A records, Request AAAA records
  reach    : 0x00000000 (Not Reachable)
  order    : 300200
resolver #4
  domain   : 8.e.f.ip6.arpa
  options  : mdns
  timeout  : 5
  flags    : Request A records, Request AAAA records
  reach    : 0x00000000 (Not Reachable)
  order    : 300400
resolver #5
  domain   : 9.e.f.ip6.arpa
  options  : mdns
  timeout  : 5
  flags    : Request A records, Request AAAA records
  reach    : 0x00000000 (Not Reachable)
  order    : 300600
resolver #6
  domain   : a.e.f.ip6.arpa
  options  : mdns
  timeout  : 5
  flags    : Request A records, Request AAAA records
  reach    : 0x00000000 (Not Reachable)
  order    : 300800
resolver #7
  domain   : b.e.f.ip6.arpa
  options  : mdns
  timeout  : 5
  flags    : Request A records, Request AAAA records
  reach    : 0x00000000 (Not Reachable)
  order    : 301000
DNS configuration (for scoped queries)
resolver #1
  search domain[0] : lan
  nameserver[0] : 192.168.1.254
  if_index : 7 (en0)
  flags    : Scoped, Request A records
  reach    : 0x00020002 (Reachable,Directly Reachable Address)
```

**Comment by smammy** [comment]
This was a _huge_ hassle to figure out, so I wrote up a little guide in hopes that others would find it helpful:

# How to convince macOS to do IPv6 DNS lookups when your only IPv6 address is via a VPN or tunnel of some sort

## The Problem

macOS's domain name resolver will only return IPv6 addresses (from AAAA records) when it thinks that you have a valid routable IPv6 address. For physical interfaces like Ethernet or Wi-Fi it's enough to set or be assigned an IPv6 address, but for tunnels (such as those using `utun` interfaces) there are some extra annoying steps that need to be taken to convince the system that yes, you indeed have an IPv6 address, and yes, you'd like to get IPv6 addresses back for DNS lookups.

I use `wg-quick` to establish a WireGuard tunnel between my laptop and a Linode virtual server. WireGuard uses a `utun` user-space tunnel device to make the connection. Here's how that device gets configured:

    utun1: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 1420
        inet 10.75.131.2 --> 10.75.131.2 netmask 0xffffff00
        inet6 fe80::a65e:60ff:fee1:b1bf%utun1 prefixlen 64 scopeid 0xc
        inet6 2600:3c03::de:d002 prefixlen 116
        nd6 options=201<PERFORMNUD,DAD>

And here's a few relevant lines from my routing table:

    Internet:
    Destination        Gateway            Flags        Refs      Use   Netif Expire
    0/1                utun1              USc             0        0   utun1
    default            10.20.4.4          UGSc            0        0     en3
    10.20.4/24         link#14            UCS             3        0     en3      !
    10.75.131.2        10.75.131.2        UH              0        0   utun1
    50.116.51.30       10.20.4.4          UGHS            7  2629464     en3
    128.0/1            utun1              USc             5        0   utun1

    Internet6:
    Destination                             Gateway                         Flags         Netif Expire
    ::/1                                    utun1                           USc           utun1
    2600:3c03::de:d000/116                  fe80::a65e:60ff:fee1:b1bf%utun1 Uc            utun1
    8000::/1                                utun1                           USc           utun1

* `10.20.4/24` is my local ethernet network.
* `10.20.4.5` is my laptop's LAN IP address.
* `10.20.4.4` is my gateway's LAN IP address.
* `10.75.131.2` is the IPv4 address of my end of the WireGuard point-to-point tunnel.
* `2600:3c03::de:d002` is the IPv6 address of my end of the WireGuard point-to-point tunnel.
* `50.116.51.30` is the public address of my Linode server.

This should be enough to have IPv6 connectivity, right? Well, name resolution works when `host` talks directly to my name server:

    sam@shiny ~> host ipv6.whatismyv6.com
    ipv6.whatismyv6.com has IPv6 address 2607:f0d0:3802:84::128

Pinging by IPv6 address works:

    sam@shiny ~> ping6 -c1 2607:f0d0:3802:84::128
    PING6(56=40+8+8 bytes) 2600:3c03::de:d002 --> 2607:f0d0:3802:84::128
    16 bytes from 2607:f0d0:3802:84::128, icmp_seq=0 hlim=55 time=80.991 ms

    --- 2607:f0d0:3802:84::128 ping6 statistics ---
    1 packets transmitted, 1 packets received, 0.0% packet loss
    round-trip min/avg/max/std-dev = 80.991/80.991/80.991/0.000 ms

And HTTP connections by IPv6 address work:

    sam@shiny ~> curl -s 'http://[2607:f0d0:3802:84::128]' -H 'Host: ipv6.whatismyv6.com' | html2text | head -3
                     This page shows your IPv6 and/or IPv4 address
                              You are connecting with an IPv6 Address of:
                                                 2600:3c03::de:d002

However, HTTP connections by IPv6-only hostname don't work:

    sam@shiny ~> curl 'http://ipv6.whatismyv6.com'
    curl: (6) Could not resolve host: ipv6.whatismyv6.com

The result is the same in `wget` as well as in GUI apps like Firefox: connecting by a literal IPv6 address works fine, but connecting by a hostname that only has an AAAA record (and no A record) associated with it does not.

Interestingly, `ping6` _is_ able to do a DNS lookup and get an IPv6 address back:

    sam@shiny ~ [6]> ping6 -c1 ipv6.whatismyv6.com
    PING6(56=40+8+8 bytes) 2600:3c03::de:d002 --> 2607:f0d0:3802:84::128
    16 bytes from 2607:f0d0:3802:84::128, icmp_seq=0 hlim=55 time=49.513 ms

    --- ipv6.whatismyv6.com ping6 statistics ---
    1 packets transmitted, 1 packets received, 0.0% packet loss
    round-trip min/avg/max/std-dev = 49.513/49.513/49.513/0.000 ms

Why can `ping6` do this when nothing else can? It turns out that when `ping6` calls `getaddrinfo` it overwrites the default flags. One of the default flags is `AI_ADDRCONFIG`, which tells the resolver to only return addresses in address families that the system has an IP address for. (That is, don't return IPv6 addresses unless the system has a (not link-local) IPv6 address.) Most other programs _add_ to the default flags rather than clobbering them, which I suppose is sensible.

If you run `scutil --dns` it will tell you how the resolver is set up. Here's the output on my system (minus a bunch of mdns stuff that doesn't matter):

    DNS configuration

    resolver #1
      search domain[0] : home.munkynet.org
      nameserver[0] : 10.20.4.4
      if_index : 14 (en3)
      flags    : Request A records
      reach    : 0x00020002 (Reachable,Directly Reachable Address)

    DNS configuration (for scoped queries)

    resolver #1
      search domain[0] : home.munkynet.org
      nameserver[0] : 10.20.4.4
      if_index : 14 (en3)
      flags    : Scoped, Request A records
      reach    : 0x00020002 (Reachable,Directly Reachable Address)

Note that under `flags`, it says `Request A records` but not `Request AAAA records`. So it's left to us to try to convince macOS's resolver that we do in fact have a valid IPv6 address, even though it's on a tunnel interface.

## SystemConfiguration

The "right" way for this to happen is for whatever program sets up the tunnel to use the bizarre and largely undocumented `SystemConfiguration` API to register the network "service" and its IPv6 properties. The Viscosity app does this. Tunnelblick does not, the official OpenVPN Client does not, and `wg-quick` sure as hell doesn't.

## The `scutil` Kludge

We can create the same SystemConfiguration "service" strucures manually using the `scutil` command:

First we create the IPv4 part of the service:

    sam@shiny ~> sudo scutil
    > d.init
    > d.add Addresses * 10.75.131.2
    > d.add DestAddresses * 10.75.131.2
    > d.add InterfaceName utun1
    > set State:/Network/Service/my_ipv6_tunnel_service/IPv4
    > set Setup:/Network/Service/my_ipv6_tunnel_service/IPv4

And then we create the IPv6 part:

    > d.init
    > d.add Addresses * fe80::a65e:60ff:fee1:b1bf 2600:3c03::de:d002
    > d.add DestAddresses * ::ffff:ffff:ffff:ffff:0:0 ::
    > d.add Flags * 0 0
    > d.add InterfaceName utun1
    > d.add PrefixLength * 64 116
    > set State:/Network/Service/my_ipv6_tunnel_service/IPv6
    > set Setup:/Network/Service/my_ipv6_tunnel_service/IPv6
    > quit

Once this is done, the output of `scutil --dns` (again modulo mdns stuff) changes:

    DNS configuration

    resolver #1
      search domain[0] : home.munkynet.org
      nameserver[0] : 10.20.4.4
      if_index : 14 (en3)
      flags    : Request A records, Request AAAA records
      reach    : 0x00020002 (Reachable,Directly Reachable Address)

    DNS configuration (for scoped queries)

    resolver #1
      search domain[0] : home.munkynet.org
      nameserver[0] : 10.20.4.4
      if_index : 14 (en3)
      flags    : Scoped, Request A records
      reach    : 0x00020002 (Reachable,Directly Reachable Address)

Now we see `Request AAAA records` in the flags! I'm not really sure what "scoped queries" are or why the DNS configuration for them didn't change, but things seem to work now so whatever:

    sam@shiny ~> curl -s 'http://ipv6.whatismyv6.com' | html2text | head -3
                     This page shows your IPv6 and/or IPv4 address
                              You are connecting with an IPv6 Address of:
                                                 2600:3c03::de:d002

When disconnecting from the tunnel, all you have to do is remove the SystemConfiguration keys you added:

    sam@shiny ~> sudo scutil
    > remove State:/Network/Service/my_ipv6_tunnel_service/IPv4
    > remove Setup:/Network/Service/my_ipv6_tunnel_service/IPv4
    > remove State:/Network/Service/my_ipv6_tunnel_service/IPv6
    > remove Setup:/Network/Service/my_ipv6_tunnel_service/IPv6
    > quit

A couple things to note:

* The name `my_ipv6_tunnel_service` is totally arbitrary.
* According to information I gleaned from the up/down scripts in the Mullvad `.ovpn` profile, you have to create both the `Setup:` and `State:` keys. I didn't verify this because I am lazy.
* I have no clue where the IPv6 `DestAddresses` come from. I copied these from Viscosity because they seemed to work there. `::ffff:ffff:ffff:ffff:0:0` for the link-local address and `::` for the public 
* I don't even really know what `DestAddresses` means or what it's used for.

## A nice script

I wrote a python script that gleans addresses and prefix lengths from `ifconfig` output. It requires Python 3.6 or later so make sure you've got that in your path. It's called `wg-updown` and calls its SystemConfiguration service `wg-updown-utun#`, but it's not really WireGuard-specific. You could call it as a post-up/pre-down script for any old VPN tunnel or run it manually. Call it like this:

    # After tunnel comes up
    wg-updown up IFACE
    
    # Before tunnel goes down
    wg-updown down IFACE

replace `IFACE` with the name of the interface that your tunnel/VPN client is using, e.g. `utun1`. It will print the commands that it's sending to `scutil` so you can see what it's doing in detail.

    #!/usr/bin/env python3

    import re
    import subprocess
    import sys

    def service_name_for_interface(interface):
        return 'wg-updown-' + interface

    v4pat = re.compile(r'^\s*inet\s+(\S+)\s+-->\s+(\S+)\s+netmask\s+\S+')
    v6pat = re.compile(r'^\s*inet6\s+(\S+?)(?:%\S+)?\s+prefixlen\s+(\S+)')
    def get_tunnel_info(interface):
        ipv4s = dict(Addresses=[], DestAddresses=[])
        ipv6s = dict(Addresses=[], DestAddresses=[], Flags=[], PrefixLength=[])
        ifconfig = subprocess.run(["ifconfig", interface], capture_output=True,
                                  check=True, text=True)
        for line in ifconfig.stdout.splitlines():
            v6match = v6pat.match(line)
            if v6match:
                ipv6s['Addresses'].append(v6match[1])
                # This is cribbed from Viscosity and probably wrong.
                if v6match[1].startswith('fe80'):
                    ipv6s['DestAddresses'].append('::ffff:ffff:ffff:ffff:0:0')
                else:
                    ipv6s['DestAddresses'].append('::')
                ipv6s['Flags'].append('0')
                ipv6s['PrefixLength'].append(v6match[2])
                continue
            v4match = v4pat.match(line)
            if v4match:
                ipv4s['Addresses'].append(v4match[1])
                ipv4s['DestAddresses'].append(v4match[2])
                continue
        return (ipv4s, ipv6s)

    def run_scutil(commands):
        print(commands)
        subprocess.run(['scutil'], input=commands, check=True, text=True)

    def up(interface):
        service_name = service_name_for_interface(interface)
        (ipv4s, ipv6s) = get_tunnel_info(interface)
        run_scutil('\n'.join([
            f"d.init",
            f"d.add Addresses * {' '.join(ipv4s['Addresses'])}",
            f"d.add DestAddresses * {' '.join(ipv4s['DestAddresses'])}",
            f"d.add InterfaceName {interface}",
            f"set State:/Network/Service/{service_name}/IPv4",
            f"set Setup:/Network/Service/{service_name}/IPv4",
            f"d.init",
            f"d.add Addresses * {' '.join(ipv6s['Addresses'])}",
            f"d.add DestAddresses * {' '.join(ipv6s['DestAddresses'])}",
            f"d.add Flags * {' '.join(ipv6s['Flags'])}",
            f"d.add InterfaceName {interface}",
            f"d.add PrefixLength * {' '.join(ipv6s['PrefixLength'])}",
            f"set State:/Network/Service/{service_name}/IPv6",
            f"set Setup:/Network/Service/{service_name}/IPv6",
        ]))

    def down(interface):
        service_name = service_name_for_interface(interface)
        run_scutil('\n'.join([
            f"remove State:/Network/Service/{service_name}/IPv4",
            f"remove Setup:/Network/Service/{service_name}/IPv4",
            f"remove State:/Network/Service/{service_name}/IPv6",
            f"remove Setup:/Network/Service/{service_name}/IPv6",
        ]))

    def main():
        operation = sys.argv[1]
        interface = sys.argv[2]
        if operation == 'up':
            up(interface)
        elif operation == 'down':
            down(interface)
        else:
            raise NotImplementedError()

    if __name__ == "__main__":
        main()

**Comment by SimonClifford** [comment]
Sam, thanks for all this work, and the nice write-up.  We're trying to roll out something to our users to workaround this issue and we want it to be as simple as possible.  To this end I've made some modifications to your python script that might be useful to others.  First, the default version of python that ships with the current (10.14 Mojave) macOS is 2.7.10, so I've backported the script to that version of python.  Second, we're making use of Tunnelblick's  (or OpenVPN's) script calling feature, so I've made the script accept its interface argument via an environment variable.

Calling the script from Tunnelblick (or OpenVPN presumably) is done by putting the following in the `.ovpn` file:

```
route-up "wg-updown.py up"
route-down "wg-updown.py down"
```

You can make this into a Tunnelblick config "file" by copying the python script and config file into a directory, then appending `.tblk` to the directory's name.  Then just drag the "file" onto Tunnelblick to import.  Note that Tunnelblick will quite rightly put up horrifying warnings if you import a config with such lines in!

```
#!/usr/bin/env python

from __future__ import print_function
import re
import subprocess
import sys
import os

def service_name_for_interface(interface):
    return 'wg-updown-' + interface

v4pat = re.compile(r'^\s*inet\s+(\S+)\s+-->\s+(\S+)\s+netmask\s+\S+')
v6pat = re.compile(r'^\s*inet6\s+(\S+?)(?:%\S+)?\s+prefixlen\s+(\S+)')
def get_tunnel_info(interface):
    ipv4s = dict(Addresses=[], DestAddresses=[])
    ipv6s = dict(Addresses=[], DestAddresses=[], Flags=[], PrefixLength=[])
    ifconfig = subprocess.check_output(["/sbin/ifconfig", interface],
                                       universal_newlines=True)
    for line in ifconfig.splitlines():
        print (line, file=sys.stderr)
        v6match = v6pat.match(line)
        if v6match:
            ipv6s['Addresses'].append(v6match.group(1))
            # This is cribbed from Viscosity and probably wrong.
            if v6match.group(1).startswith('fe80'):
                ipv6s['DestAddresses'].append('::ffff:ffff:ffff:ffff:0:0')
            else:
                ipv6s['DestAddresses'].append('::')
            ipv6s['Flags'].append('0')
            ipv6s['PrefixLength'].append(v6match.group(2))
            continue
        v4match = v4pat.match(line)
        if v4match:
            ipv4s['Addresses'].append(v4match.group(1))
            ipv4s['DestAddresses'].append(v4match.group(2))
            continue
    return (ipv4s, ipv6s)

def run_scutil(commands):
    print(commands)
    p = subprocess.Popen(['/usr/sbin/scutil'], stdin=subprocess.PIPE)
    p.communicate(commands.encode('utf-8'))
    retval = p.wait()
    if retval != 0:
        raise subprocess.OSError('Error: scutil returned ' + retval)

def up(interface):
    service_name = service_name_for_interface(interface)
    (ipv4s, ipv6s) = get_tunnel_info(interface)
    run_scutil('\n'.join([
        "d.init",
        "d.add Addresses * {}".format(' '.join(ipv4s['Addresses'])),
        "d.add DestAddresses * {}".format(' '.join(ipv4s['DestAddresses'])),
        "d.add InterfaceName {}".format(interface),
        "set State:/Network/Service/{}/IPv4".format(service_name),
        "set Setup:/Network/Service/{}/IPv4".format(service_name),
        "d.init",
        "d.add Addresses * {}".format(' '.join(ipv6s['Addresses'])),
        "d.add DestAddresses * {}".format(' '.join(ipv6s['DestAddresses'])),
        "d.add Flags * {}".format(' '.join(ipv6s['Flags'])),
        "d.add InterfaceName {}".format(interface),
        "d.add PrefixLength * {}".format(' '.join(ipv6s['PrefixLength'])),
        "set State:/Network/Service/{}/IPv6".format(service_name),
        "set Setup:/Network/Service/{}/IPv6".format(service_name),
    ]))

def down(interface):
    service_name = service_name_for_interface(interface)
    run_scutil('\n'.join([
        "remove State:/Network/Service/{}/IPv4".format(service_name),
        "remove Setup:/Network/Service/{}/IPv4".format(service_name),
        "remove State:/Network/Service/{}/IPv6".format(service_name),
        "remove Setup:/Network/Service/{}/IPv6".format(service_name),
    ]))

def main():
    operation = sys.argv[1]
#    interface = sys.argv[2]
    interface = os.environ['dev']
    if operation == 'up':
        up(interface)
    elif operation == 'down':
        down(interface)
    else:
        raise NotImplementedError()

if __name__ == "__main__":
    main()
```

If the Tunnelblick authors could integrate this functionality natively, we'd be very grateful.

**Comment by janmuennich** [comment]
As an easier workaround I added a second network location that I use in IPv4-only environments. For this location I have configured fake (valid but not routable) manual IPv6 address and router. This doesn't seem to have any side effects but gives working IPv6 after OpenVPN is connected.

**Comment by bert128** [comment]
Any chance one of the scripts referenced above can be merged into client.up.tunnelblick.sh? Currently i modify this manually, but it triggers a warning that the software has been modified and gets overwritten on update...

**Comment by bert128** [comment]
Note that if you have an IPv6-only VPN with or without NAT64, then this bug prevents it from working at all if the native connection lacks IPv6. The same connection works fine on iOS (openvpn connect) and on android.

**Comment by jkbullard** [comment]
@bert128  wrote
> Any chance one of the scripts referenced above can be merged into client.up.tunnelblick.sh? Currently i modify this manually, but it triggers a warning that the software has been modified and gets overwritten on update...

 1. We wouldn't incorporate this into Tunnelblick unless it was rewritten as a bash version 3 script. (We use only bash scripts, to avoid problems with different versions of macOS including different versions of python and other tools, and to avoid adding an additional dependency to Tunnelblick.)

 2. You could probably use an `up-prefix.sh` and/or `up-suffix.sh` script instead of modifying Tunnelblick's `client.up.tunnelblick.sh` script. See [Using Scripts](https://tunnelblick.net/cUsingScripts.html) for details.

**Comment by bert128** [comment]
The version linked here: https://github.com/eduvpn/macos/pull/221 is written in bash.
I currently have a very crufty workaround, but it would be much better if the default install handled IPv6 properly. Having to provide scripts to users and explain the resulting security warnings is painful.

**Comment by fxck** [comment]
I'm experiencing the same problem @bert128, after Ventura update the workaround @smammy [posted](https://gist.github.com/smammy/3247b5114d717d12b68c201000ab163d) stopped working (`Request AAAA records`) not showing up anymore. Have you found a different workaround?

**Comment by bebehei** [comment]
Can anyone give a hint, whether the proposed scripts still work on macOS Ventura (or Sonoma)? I cannot get the python script nor the bash equivalent from eduvpn/macos#221 to work on Ventura. (Sonoma update outstanding, will do, if it might fix)

The "give your device a bogus IPv6 address" still works for me. Only thing: If you're not in the VPN and actually have a working IPv6 connection, you don't have IPv6 connectivity and need to set it back to "Configure IPv6: Automatically".

I'm not a macOS pro (using macOS now for a few months). For me, this seems like a bug inside macOS's IPv6 reachability calculation.

I have dumped all `scutil` keys in the different states (connected: yes/no, faked ipv6: yes/no).

The outstanding records for me, are these:

```text
[benediktheine:~] % scutil
> show State:/Network/Global/IPv6
<dictionary> {
  PrimaryInterface : en0
  PrimaryService : 587A4CBF-D3B7-4F9A-A332-2FFD22A541AC
  Router : fd::
}
> show State:/Network/Global/SMB
<dictionary> {
  NetBIOSNodeType : Mixed
}
> 
```

Thinking about this (with my unaquainted mind), I do think, the key `State:/Network/Global/IPv6` is completely missing and this hinders some other (probably macOS-related) algorithm, to disable resolution of AAAA-Records.

(While searching for `State:/Network/Global/IPv6`, the only google match had been #116)

**Comment by jkbullard** [comment]
To @bebehei, @fxck, @bert128, @janmuennich, @SimonClifford, @smammy, @mj-harvey, @wido (or anyone else experiencing this Issue):

Please post the Diagnostic Info obtained from following the directions in [Read Before You Post](https://tunnelblick.net/cBeforeYouPost.html) as an attachment to a comment.

**Comment by bert128** [comment]
Logs attached as requested, connecting from a legacy network to a dual stack VPN:

[scutil-1.txt](https://github.com/Tunnelblick/Tunnelblick/files/14297222/scutil-1.txt)
[diagnostic.txt](https://github.com/Tunnelblick/Tunnelblick/files/14297224/diagnostic.txt)
[scutil-2.txt](https://github.com/Tunnelblick/Tunnelblick/files/14297225/scutil-2.txt)

Also output from scutil both before and after connecting.

You can also see that the server pushes some IPv6 nameservers, but they don't get used and a warning is given:

```
2024-02-15 13:03:25.684524 *Tunnelblick: Warning: DNS server address 2a07:a880:4601:10f0:cd45::1 is not being used.
2024-02-15 13:03:25.692462 *Tunnelblick: Warning: DNS server address 2001:67c:750:1:cafe:cd45::1 is not being used.
```

**Comment by bert128** [comment]
The problem also happens the other way, if you are on an IPv6-only network and your DNS is set to only request AAAA records but you connect to a dualstack VPN it doesn't add requesting A records.

Something else i noticed is that the DNS entry reported by scutil still reports the interface used for DNS as "en0", and after a while reports the nameservers as "not reachable":
```
resolver #1
  search domain[0] : openvpn
  nameserver[0] : 192.165.9.158
  nameserver[1] : 46.227.67.134
  if_index : 12 (en0)
  flags    : Scoped, Request A records
  reach    : 0x00000000 (Not Reachable)
```

It's trying to use en0 to reach the DNS servers, which does not work as the traffic is now routed through the tunnel interface. After a while it reports the nameservers as not reachable, and DNS stops working entirely. If you manually add a route to the nameservers via en0 and the nameservers allow the traffic (ie they're public and not restricted to the vpn) then DNS continues working.

When using Cisco Anyconnect everything works as it should - the resolvers pushed by the VPN are correctly routed via the utun* interface and it also sets the A/AAAA options according to the tunnel it connected to.

**Comment by bert128** [comment]
When using AnyConnect, instead of trying to overwrite the parameters of resolver 1, a second resolver entry is created:

```
resolver #1
  nameserver[0] : XXX:XXX
  nameserver[1] : XXX:XXX
  if_index : 6 (en0)
  flags    : Scoped, Request AAAA records
  reach    : 0x00000002 (Reachable)
 
resolver #2
  search domain[0] : vpn.internal
  nameserver[0] : XXX:XXX
  nameserver[1] : XXX:XXX
  if_index : 19 (utun4)
  flags    : Scoped, Request A records, Request AAAA records
  reach    : 0x00000002 (Reachable)
  order    : 1
```

The second entry has the correct DNS flags for the VPN being connected to, irrespective of what the original interface was set to. It also sets if_index to its tunnel interface. 
The second one takes precedence, the first one is ignored while the VPN is connected.

**Comment by durkovic** [comment]
On MacOS Ventura, it's necessary to inform the OS about IPv6 interface - then all resolvers will automatically start requesting AAAA records. Here is a sample script:

```
#!/bin/sh
scutil << EOF
d.init
d.add Addresses * your:ipv6:address
d.add ConfirmedInterfaceName gif0
d.add InterfaceName gif0
d.add NetworkSignature IPv6.6to4=(IPv4.Router=ipv4.router.address;IPv4.RouterHardwareAddress=53:12:45:ca:fe:00)
d.add PrefixLength # 64
d.add Router router:ipv6:address
set State:/Network/Service/291DF5CE-6E23-4C71-805F-9CFDCCDEC795/IPv6
EOF
```

scutil --nwi should show reachable IPv6 network interface (gif0).
scutil --dns should show Request A records, Request AAAA records for all items except the IPv4-only interface.


After VPN disconnect, remove that item:

```
#!/bin/sh
scutil << EOF
remove State:/Network/Service/291DF5CE-6E23-4C71-805F-9CFDCCDEC795/IPv6
EOF
```

**Comment by altf4arnold** [comment]
It surprises me to see that this is still a problem. 

Have just been trying, if I set up ipv6 manually on my WiFi interface (Mac book air M1) on an IPv4 only network using a unrouted prefix (tried with 2001:db8::/32 for proof of concept), sudently, scutil --dns goes with both records. Kind of a dirty patch, but works on Sonoma.

**Comment by jkbullard** [comment]
@ altf4arnold  wrote:
> It surprises me to see that this is still a problem.

It will be a problem until someone who cares about it submits a fix as a [Pull Request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests). That's what [Free Software](https://www.fsf.org/about/what-is-free-software) is all about!

**Comment by altf4arnold** [comment]
@jkbullard I've been looking around in the codebase for a little while, there doesn't seem to be an elegant solution on the current release of mac os. (I might be wrong. but at this point, I don't see much I can add)

**Comment by p1kus** [comment]
This helped me diagnose and solve issues with Cisco Secure Client dropping connections to Microsoft domains, thank you!
